lib/ansible/utils/encrypt.py

```diff
@@ -33,31 +40,140 @@
 DEFAULT_PASSWORD_LENGTH = 20
 
 
-def do_encrypt(result, encrypt, salt_size=None, salt=None):
-    if PASSLIB_AVAILABLE:
+class BaseHash(object):
+    algo = namedtuple('algo', ['crypt_id', 'salt_size', 'implicit_rounds'])
+    algorithms = {
+        'md5_crypt': algo(crypt_id='1', salt_size=8, implicit_rounds=None),
+        'bcrypt': algo(crypt_id='2a', salt_size=22, implicit_rounds=None),
+        'sha256_crypt': algo(crypt_id='5', salt_size=16, implicit_rounds=5000),
+        'sha512_crypt': algo(crypt_id='6', salt_size=16, implicit_rounds=5000),
+    }
+
+    def __init__(self, algorithm, error_type):
+        self.algorithm = algorithm
+        self.error_type = error_type
+
+
+class CryptHash(BaseHash):
+    def __init__(self, algorithm, error_type):
+        super(CryptHash, self).__init__(algorithm, error_type)
+
+        if sys.platform.startswith('darwin'):
+            raise self.error_type("crypt.crypt not supported on Mac OS X/Darwin, install passlib python module")
+
+        if algorithm not in self.algorithms:
+            raise self.error_type("crypt.crypt does not support '%s' algorithm" % self.algorithm)
+        self.algo_data = self.algorithms[algorithm]
+
+    def hash(self, secret, **settings):
```

Comment:
Note: organizationally, our code is largely, callers near the bottom of the file, callees near the top.  (Like C code).